File: Program's_Contributed_By_Contributors/Python_Programs/merge_sort.py
def merge(a):
    if len(a)>1:
        mid=(len(a))//2
        l=a[:mid]
        r=a[mid:]
        merge(l)
        merge(r)
        i=j=k=0
        while i<len(l) and j<len(r):
            if l[i]<r[j]:
                a[k]=l[i]
                i+=1
            else:
                a[k]=r[j]
                j+=1
            k+=1
        while i<len(l):
            a[k]=l[i]
            i+=1
            k+=1
        while j<len(r):
            a[k]=r[j]
            j+=1
            k+=1
def print_sortedarray(a):
    for i in range(len(a)):
        print(a[i],end=" ")
a=[5,6,8,0,3,4,21,1]
merge(a)
print_sortedarray(a)

File: Program's_Contributed_By_Contributors/C++ Programs/Sorting/MergeSort.cpp
#include<iostream>
#include<vector>
#define MAX 100000
using namespace std;

//You are given two sorted arrays A and B. You want to merge these arrays and return a sorted vector will elements from both arrays.
void merge(vector<int> &a, vector<int> &b, vector<int> &toreturn)
{ 
    int apointer = 0, bpointer = 0; //point to the first element in the vectors
	
    //That hasn't been taken yet
    while(apointer<a.size() && bpointer<b.size())
	{
        if(a[apointer]<=b[bpointer]){
            toreturn.push_back(a[apointer++]);
        }
        else{ //b[bpointer] < a[apointer]
//            inv += (a.size() - apointer);
            toreturn.push_back(b[bpointer++]);
        }
    }
    while(apointer<a.size()) toreturn.push_back(a[apointer++]);
    while(bpointer<b.size()) toreturn.push_back(b[bpointer++]);
}

void ms(int arr[], int s, int e){
//	cout<<s<<" "<<e<<endl;
	if(s==e){// there is only one or no element
		return;
	}
	int mid = (s+e)/2;
	ms(arr, s, mid);
	ms(arr, mid+1, e);
	
	// Now arr[] has two sorted parts inside it. 
	// We have to take them out, merge the two parts and 
	// then resassign the values to arr[]
	vector<int> a,b;
	for(int i=s; i<=mid; i++) a.push_back(arr[i]);
	for(int i=mid+1; i<=e; i++) b.push_back(arr[i]);  
	
	vector<int> temp;
	merge(a,b, temp);
	for(int i=0; i<temp.size(); i++)
		arr[s+i] = temp[i];
	
}



int main()
{
	int n; 
	cin>>n; 
	int arr[n];
	for(int i=0; i<n; i++) cin>>arr[i];
	ms(arr, 0, n-1);
	for(int i=0; i<n; i++) cout<<arr[i]<<" ";
	cout<<endl;
}